renormalizer/mps/mpo.py

```diff
@@ -2007,18 +2017,49 @@ def apply(self, mp: MatrixProduct, canonicalise: bool=False) -> MatrixProduct:
             new_mps.canonicalise()
         return new_mps
 
-    def contract(self, mps, check_emtpy=False):
+    def contract(self, mps, check_emtpy=False, algo="svd"):
+        r""" an approximation of mpo @ mps/mpdm/mpo
+        
+        Parameters
+        ----------
+        mps : renormalizer.mps.Mps, renormalizer.mps.Mpo, renormalizer.mps.MpDm,
+        check_emtpy : bool, optional
+            Check if the obtained new mps has zero local matrix. Default is
+            ``False``.
+        algo: str, optional
+            The algorithm to compress mpo @ mps/mpdm/mpo.  It could be `svd`
+            (default) and `variational`. 
+        
+        Returns
+        -------
+        new_mps : an approximation of mpo @ mps/mpdm/mpo. The original `mps` is
+            not overwritten.
+
+        See Also
+        --------
+        renormalizer.mps.mp.MatrixProduct.compress
+        renormalizer.mps.mp.MatrixProduct.variational_compress
+
 
         """
-        mapply->canonicalise->compress
-        """
-        new_mps = self.apply(mps)
-        if check_emtpy:
-            for mt in new_mps:
-                if mt.nearly_zero():
-                    raise EmptyMatrixError
-        new_mps.canonicalise()
-        new_mps.compress()
+        if algo == "svd":
+            # mapply->canonicalise->compress
+            new_mps = self.apply(mps)
+            if check_emtpy:
+                for mt in new_mps:
+                    if mt.nearly_zero():
+                        raise EmptyMatrixError
```

Comment:
If I remember correctly, the `EmptyMatrixError` is the legacy of the days when `scheme4` has no zero exciton state. I'll prepare a PR to remove this feature after this PR is merged.

Reply:
Ok